diff -u a/config.h.in b/config.h.in
--- a/config.h.in	2018-10-20 03:20:46.000000000 +0700
+++ b/config.h.in	2019-08-24 17:35:28.138615962 +0700
@@ -441,3 +441,8 @@
 
 /* Define to __typeof__ if your compiler spells it that way. */
 #undef typeof
+
+/* Lilith */
+#ifdef __lilith__
+#define CAIRO_NO_MUTEX 1
+#endif
diff -u a/build/config.sub b/build/config.sub
--- a/build/config.sub	2018-08-17 08:11:52.000000000 +0700
+++ b/build/config.sub	2019-08-24 18:54:07.789484403 +0700
@@ -1363,7 +1363,7 @@
 	# The portable systems comes first.
 	# Each alternative MUST end in a * to match a version number.
 	# -sysv* is not here because it comes later, after sysvr4.
-	-gnu* | -bsd* | -mach* | -minix* | -genix* | -ultrix* | -irix* \
+	-gnu* | -bsd* | -mach* | -minix* | -genix* | -ultrix* | -irix* | -lilith* \
 	      | -*vms* | -sco* | -esix* | -isc* | -aix* | -cnk* | -sunos | -sunos[34]*\
 	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
 	      | -sym* | -kopensolaris* | -plan9* \
diff -u a/src/cairo-path-stroke-polygon.c b/src/cairo-path-stroke-polygon.c
--- a/src/cairo-path-stroke-polygon.c	2018-08-17 08:10:53.000000000 +0700
+++ b/src/cairo-path-stroke-polygon.c	2019-09-21 16:09:20.804159623 +0700
@@ -820,7 +820,7 @@
     add_cap (stroker, face, c);
 }
 
-static inline double
+static double __attribute__((optimize("O0")))
 normalize_slope (double *dx, double *dy)
 {
     double dx0 = *dx, dy0 = *dy;
diff -u a/src/cairo-tor-scan-converter.c b/src/cairo-tor-scan-converter.c
--- a/src/cairo-tor-scan-converter.c	2018-10-17 00:03:07.000000000 +0700
+++ b/src/cairo-tor-scan-converter.c	2019-09-21 16:09:10.193896097 +0700
@@ -791,7 +791,7 @@
  * This function depends on being called with all edges from the
  * active list in the order they appear on the list (i.e. with
  * non-decreasing x-coordinate.)  */
-static void
+static void __attribute__((optimize("O0")))
 cell_list_render_edge(struct cell_list *cells,
 		      struct edge *edge,
 		      int sign)
@@ -1462,7 +1462,7 @@
     (out) = tmp__;						\
 } while (0)
 
-inline static void
+inline static void __attribute__((optimize("00")))
 polygon_add_edge (struct polygon *polygon,
 		  const cairo_edge_t *edge)
 {
@@ -1713,7 +1713,7 @@
 }
 
 
-I void
+I void __attribute__((optimize("O0")))
 glitter_scan_converter_render(glitter_scan_converter_t *converter,
 			      unsigned int winding_mask,
 			      int antialias,
