diff --git a/src/compiler/crystal/compiler.cr b/src/compiler/crystal/compiler.cr
index 3853d8962..ca5918acd 100644
--- a/src/compiler/crystal/compiler.cr
+++ b/src/compiler/crystal/compiler.cr
@@ -310,7 +310,7 @@ module Crystal
         unit.emit(emit, emit_base_filename || output_filename)
       end
 
-      target_machine.emit_obj_to_file llvm_mod, object_name
+      # target_machine.emit_obj_to_file llvm_mod, object_name
 
       stdout.puts linker_command(program, object_name, output_filename, nil)
     end
diff --git a/src/empty.cr b/src/empty.cr
index 8a78751e0..c68218f07 100644
--- a/src/empty.cr
+++ b/src/empty.cr
@@ -1,11 +1 @@
 require "primitives"
-
-lib LibCrystalMain
-  @[Raises]
-  fun __crystal_main(argc : Int32, argv : UInt8**)
-end
-
-fun main(argc : Int32, argv : UInt8**) : Int32
-  LibCrystalMain.__crystal_main(argc, argv)
-  0
-end
